# Testing Notebook for ETL

Imports functions from oas_onboarding/etl and tests those functions

In [1]:
import sys
import pandas as pd
from Bio.Align import PairwiseAligner

sys.path.insert(0, '/home/cameronhu/oas_onboarding/src/etl')
print(sys.path)

['/home/cameronhu/oas_onboarding/src/etl', '/home/cameronhu/miniforge3/lib/python312.zip', '/home/cameronhu/miniforge3/lib/python3.12', '/home/cameronhu/miniforge3/lib/python3.12/lib-dynload', '', '/home/cameronhu/miniforge3/lib/python3.12/site-packages']


In [2]:
from process import *

pd.set_option('display.max_columns', None)

# Unpaired Data

In [3]:
heavy_data_file = '/export/share/cameronhu/oas/unpaired/unpaired_human/unpaired_human_heavy/1279049_1_Heavy_Bulk.csv.gz'
heavy_metadata = parse_metadata(heavy_data_file)

heavy_metadata

{'Run': 1279049,
 'Link': 'https://doi.org/10.1038/s41586-022-05371-z',
 'Author': 'Jaffe et al., 2022',
 'Species': 'human',
 'Age': 35,
 'BSource': 'PBMC',
 'BType': 'Naive-B-Cells',
 'Vaccine': 'None',
 'Disease': 'SARS-COV-2',
 'Subject': 'Donor-2',
 'Longitudinal': 'no',
 'Unique sequences': 2,
 'Total sequences': 2,
 'Isotype': 'Bulk',
 'Chain': 'Heavy',
 'metadata_uid': '5db6a544-653e-4091-99c5-04ac2ed74f52'}

In [4]:
# Raw reading of OAS data file
heavy_sequence_raw = pd.read_csv(heavy_data_file, header=1)

heavy_sequence_raw.shape[1]

97

In [5]:
heavy_sequence_df, heavy_antibody_df = parse_sequence_antibody_data(heavy_data_file, heavy_metadata)
heavy_sequence_df.head()

sequence locus stop_codon  \
0  AGCTCTGAGAGAGGAGCCCAGCCCTGGGATTTTCAGGTGTTTTCAT...     H          F   
1  ATGCTTTCTGAGAGTCATGGATCTCATGTGCAAGAAAATGAAGCAC...     H          F   

  vj_in_frame v_frameshift productive rev_comp complete_vdj       v_call  \
0           T            F          T        F            T  IGHV3-23*01   
1           T            F          T        F            T  IGHV4-39*01   

        d_call    j_call                                 sequence_alignment  \
0  IGHD3-22*01  IGHJ4*02  GAGGTGCAGCTGTTGGAGTCTGGGGGAGGCTTGGTACAGCCTGGGG...   
1          NaN  IGHJ6*02  CAGCTGCAGCTGCAGGAGTCGGGCCCAGGACTGGTGAAGCCTTCGG...   

                                  germline_alignment  \
0  GAGGTGCAGCTGTTGGAGTCTGGGGGAGGCTTGGTACAGCCTGGGG...   
1  CAGCTGCAGCTGCAGGAGTCGGGCCCAGGACTGGTGAAGCCTTCGG...   

                               sequence_alignment_aa  \
0  EVQLLESGGGLVQPGGSLRLSCAASGFTFSSYAMSWVRQAPGKGLE...   
1  QLQLQESGPGLVKPSETLSLTCTVSGGSISSSSYYWGWIRQPPGKG...   

                               germline_alignment_aa  v_alignment_start  \
0  EVQLLESGGGLVQPGGSLRLSCAASGFTFSSYAMSWVRQAPGKGLE...                  1   
1  QLQLQESGPGLVKPSETLSLTCTVSGGSISSSSYYWGWIRQPPGKG...                  1   

   v_alignment_end  d_alignment_start  d_alignment_end  j_alignment_start  \
0              291              296.0            318.0                326   
1              298                NaN              NaN                310   

   j_alignment_end                               v_sequence_alignment  \
0              370  GAGGTGCAGCTGTTGGAGTCTGGGGGAGGCTTGGTACAGCCTGGGG...   
1              366  CAGCTGCAGCTGCAGGAGTCGGGCCCAGGACTGGTGAAGCCTTCGG...   

                             v_sequence_alignment_aa  \
0  EVQLLESGGGLVQPGGSLRLSCAASGFTFSSYAMSWVRQAPGKGLE...   
1  QLQLQESGPGLVKPSETLSLTCTVSGGSISSSSYYWGWIRQPPGKG...   

                                v_germline_alignment  \
0  GAGGTGCAGCTGTTGGAGTCTGGGGGAGGCTTGGTACAGCCTGGGG...   
1  CAGCTGCAGCTGCAGGAGTCGGGCCCAGGACTGGTGAAGCCTTCGG...   

                             v_germline_alignment_aa     d_sequence_alignment  \
0  EVQLLESGGGLVQPGGSLRLSCAASGFTFSSYAMSWVRQAPGKGLE...  ATTACTATGATAGTAGTGGTTAT   
1  QLQLQESGPGLVKPSETLSLTCTVSGGSISSSSYYWGWIRQPPGKG...                      NaN   

  d_sequence_alignment_aa     d_germline_alignment d_germline_alignment_aa  \
0                 YYDSSGY  ATTACTATGATAGTAGTGGTTAT                 YYDSSGY   
1                     NaN                      NaN                     NaN   

                                j_sequence_alignment j_sequence_alignment_aa  \
0      ACTTTGACTACTGGGGCCAGGGAACCCTGGTCACCGTCTCCTCAG          FDYWGQGTLVTVSS   
1  TACTACTACTACGGTATGGACGTCTGGGGCCAAGGGACCACGGTCA...     YYYYGMDVWGQGTTVTVSS   

                                j_germline_alignment j_germline_alignment_aa  \
0      ACTTTGACTACTGGGGCCAGGGAACCCTGGTCACCGTCTCCTCAG          FDYWGQGTLVTVSS   
1  TACTACTACTACGGTATGGACGTCTGGGGCCAAGGGACCACGGTCA...     YYYYGMDVWGQGTTVTVSS   

                                                fwr1  \
0  GAGGTGCAGCTGTTGGAGTCTGGGGGAGGCTTGGTACAGCCTGGGG...   
1  CAGCTGCAGCTGCAGGAGTCGGGCCCAGGACTGGTGAAGCCTTCGG...   

                     fwr1_aa                            cdr1     cdr1_aa  \
0  EVQLLESGGGLVQPGGSLRLSCAAS        GGATTCACCTTTAGCAGCTATGCC    GFTFSSYA   
1  QLQLQESGPGLVKPSETLSLTCTVS  GGTGGCTCCATCAGCAGTAGTAGTTACTAC  GGSISSSSYY   

                                                fwr2            fwr2_aa  \
0  ATGAGCTGGGTCCGCCAGGCTCCAGGGAAGGGGCTGGAGTGGGTCT...  MSWVRQAPGKGLEWVSA   
1  TGGGGCTGGATCCGCCAGCCCCCAGGGAAGGGGCTGGAGTGGATTG...  WGWIRQPPGKGLEWIGS   

                       cdr2   cdr2_aa  \
0  ATTAGTGGTAGTGGTGGTAGCACA  ISGSGGST   
1     ATCTATTATAGTGGGAGCACC   IYYSGST   

                                                fwr3  \
0  TACTACGCAGACTCCGTGAAGGGCCGGTTCACCATCTCCAGAGACA...   
1  TACTACAACCCGTCCCTCAAGAGTCGAGTCACCATATCCGTAGACA...   

                                  fwr3_aa                               fwr4  \
0  YYADSVKGRFTISRDNSKNTLYLQMNSLRAEDTAVYYC  TG

In [6]:
# The number of columns in the unpaired data after assigning UID should be 97 + 3 = 99. 
# Added columns are Antibody UID, Sequence UID, Isotype 
heavy_sequence_df.shape[1]

100

# Paired Data

In [5]:
paired_data_file = '/export/share/cameronhu/oas/paired/paired_human/1_S1__1_Paired_All.csv.gz'
paired_metadata = parse_metadata(paired_data_file)

paired_metadata

{'Run': '1_S1_',
 'Link': 'https://doi.org/10.1038/s41590-022-01230-1',
 'Author': 'Phad et al., 2022',
 'Species': 'human',
 'Age': 'no',
 'BSource': 'PBMC',
 'BType': 'Plasma-B-Cells',
 'Vaccine': 'None',
 'Disease': 'None',
 'Subject': 'Donor-2',
 'Longitudinal': 'Jul-Year-2020',
 'Unique sequences': 4668,
 'Isotype': 'All',
 'Chain': 'Paired',
 'metadata_uid': '83e9ef96-21fc-40de-a0ca-9c80370bbfc5'}

In [6]:
paired_data_raw = pd.read_csv(paired_data_file, header=1)
paired_data_raw.shape[1]

198

In [7]:
paired_df, antibody_df = parse_sequence_antibody_data(paired_data_file, paired_metadata)
paired_df.head()

sequence_id_heavy  \
0  AAACCTGAGTCAATAG-1_contig_1   
1  AAACCTGCAACAACCT-1_contig_2   
2  AAACCTGCACGTCTCT-1_contig_2   
3  AAACCTGGTCGACTAT-1_contig_2   
4  AAACCTGGTCGGCTCA-1_contig_2   

                                      sequence_heavy locus_heavy  \
0  AGCTCTGGGAGAGGAGCTCCAGCCTTGGGATTCCCAGCTGTCTCCA...           H   
1  AGCTCTGAGAGCGGAGCCCCAGCCCCAGAATTCCCAGGTGTTTTCA...           H   
2  AGCTCTGAGAGAGGAGCCCAGCCCTGGGATTTTCAGGTGTGTTCAC...           H   
3  GCTCTGAGAGAGGAGCCCAGCCCTGGGATTTTCAGGTGTTTTCATT...           H   
4  CGAGCCCAGCACTGGAAGTCGCCGGTGTTTCCATTCGGTGATCATC...           H   

  stop_codon_heavy vj_in_frame_heavy v_frameshift_heavy productive_heavy  \
0                F                 T                  F                T   
1                F                 T                  F                T   
2                F                 T                  F                T   
3                F                 T                  F                T   
4                F                 T                  F                T   

  rev_comp_heavy complete_vdj_heavy v_call_heavy d_call_heavy j_call_heavy  \
0              F                  T  IGHV3-73*02   IGHD2-2*01     IGHJ4*02   
1              F                  T  IGHV3-72*01  IGHD6-19*01     IGHJ4*02   
2              F                  T  IGHV3-23*01  IGHD6-13*01     IGHJ5*02   
3              F                  T  IGHV3-23*01  IGHD5-12*01     IGHJ4*02   
4              F                  T  IGHV3-30*04  IGHD5-12*01     IGHJ4*01   

                            sequence_alignment_heavy  \
0  GAGGTGCAGCTGGTGGAGTCCGGGGGAGGCTTGGTCCAGCCTGGGG...   
1  GAGGTGCAACTGGTGGAGTCTGGGGGAGGCTTGGTCCAGCCTGGAG...   
2  GAGGTGCAGTTGTTGGAGTCTGGGGGAGGCTTGGTACAGCCTGGGG...   
3  GAGGTGCAGCTGTTGGAGTCTGGGGGAGGCTTGGTACAGCGGGGGG...   
4  CAGGTTCACCTGGTGGAGTCTGGGGGGCGCGAGGTCCAGCCAGGGG...   

                            germline_alignment_heavy  \
0  GAGGTGCAGCTGGTGGAGTCCGGGGGAGGCTTGGTCCAGCCTGGGG...   
1  GAGGTGCAGCTGGTGGAGTCTGGGGGAGGCTTGGTCCAGCCTGGAG...   
2  GAGGTGCAGCTGTTGGAGTCTGGGGGAGGCTTGGTACAGCCTGGGG...   
3  GAGGTGCAGCTGTTGGAGTCTGGGGGAGGCTTGGTACAGCCTGGGG...   
4  CAGGTGCAGCTGGTGGAGTCTGGGGGAGGCGTGGTCCAGCCTGGGA...   

                         sequence_alignment_aa_heavy  \
0  EVQLVESGGGLVQPGGSLKLSCAASGFTFSGSAMHWVRQASGKGLE...   
1  EVQLVESGGGLVQPGGSLRLSCAASGFTFSDHYMEWVRQAPGKGLE...   
2  EVQLLESGGGLVQPGGSLRLSCAASGFTFSSVAMTWVRQAPGKGLE...   
3  EVQLLESGGGLVQRGGSLRLSCAASRFTFSSYAMTWVRQAPGKGLE...   
4  QVHLVESGGREVQPGAPLRLSCAASGFTFALSPMGWVRQAPGQGLE...   

                         germline_alignment_aa_heavy  v_alignment_start_heavy  \
0  EVQLVESGGGLVQPGGSLKLSCAASGFTFSGSAMHWVRQASGKGLE...                        1   
1  EVQLVESGGGLVQPGGSLRLSCAASGFTFSDHYMDWVRQAPGKGLE...                        1   
2  EVQLLESGGGLVQPGGSLRLSCAASGFTFSSYAMSWVRQAPGKGLE...                        1   
3  EVQLLESGGGLVQPGGSLRLSCAASGFTFSSYAMSWVRQAPGKGLE...                        1   
4  QVQLVESGGGVVQPGRSLRLSCAASGFTFSSYAMHWVRQAPGKGLE...                        1   

   v_alignment_end_heavy  d_alignment_start_heavy  d_alignment_end_heavy  \
0                    302                    303.0                  323.0   
1                    301                    310.0                  315.0   
2                    295                    297.0                  304.0   
3                    296                    299.0                  313.0   
4                    296                    305.0                  310.0   

   j_alignment_start_heavy  j_alignment_end_heavy  \
0                      332                    373   
1                      320                    361   
2                      320                    361   
3                      321                    364   
4                      313                    355   

                          v_sequence_alignment_heavy  \
0  GAGGTGCAGCTGGTGGAGTCCGGGGGAGGCTTGGTCCAGCCTGGGG...   
1  GAGGTGCAACTGGTGGAGTCTGGGGGAGGCTTGGTCCAGCCTGGAG...   
2  GAGGTGCAGTTGTTGGAGTCTGGGGGAG

In [8]:
# After processing, number of columns should be 99 (same as unpaired data frame)
len(paired_df.columns)

200

# Difference Between Paired and Unpaired Columns

There are at least two columns that don't exist in the Unpaired df that do exist in the Paired df. Raw unpaired df has 97 columns. Doubling that for the paired should yield 194 columns. However, there are 198 columns in the raw Paired df. 

One of those columns is the "Isotype" column, as that can't be stored in the metadata because there are two chains and thus two isotypes. This will be remedied as we want each sequence to also contain Isotype data.

Trying to find the missing column:

In [7]:
heavy_data_file = '/export/share/cameronhu/oas/unpaired/unpaired_human/unpaired_human_heavy/1279049_1_Heavy_Bulk.csv.gz'
paired_data_file = '/export/share/cameronhu/oas/paired/paired_human/1_S1__1_Paired_All.csv.gz'

heavy_data_raw = pd.read_csv(heavy_data_file, header=1)
paired_data_raw = pd.read_csv(paired_data_file, header=1)

def get_unique_columns_from_paired(paired_df):
    """
    Extract unique column names from Paired dataframe by removing '_heavy' and '_light' suffixes.
    """
    paired_columns = paired_df.columns
    # Remove the '_heavy' and '_light' suffixes
    unique_columns_paired = {col.replace('_heavy', '').replace('_light', '') for col in paired_columns}
    return unique_columns_paired

def compare_columns(unpaired_df, paired_df):
    """
    Compare the columns of Unpaired and Paired dataframes.
    Returns columns missing from Unpaired but present in Paired.
    """
    # Get unique columns from Unpaired
    unpaired_columns = set(unpaired_df.columns)

    # Get unique columns from Paired
    unique_columns_paired = get_unique_columns_from_paired(paired_df)

    # Find columns missing from Unpaired that exist in Paired
    missing_columns = unique_columns_paired - unpaired_columns
    return missing_columns

# Example usage
missing_columns = compare_columns(heavy_data_raw, paired_data_raw)

print("Columns missing from Unpaired but present in Paired:")
for col in missing_columns:
    print(col)


Columns missing from Unpaired but present in Paired:
sequence_id
Isotype


# Analysis of Sequence Data

Difference between 'sequence' and 'sequence_alignment' fields. I presume the first is the raw sequence from the FASTA, and the 'sequence_alignment' is the NT output from IgBLAST

In [ ]:
sequence = heavy_sequence_df['sequence'][0]
print(len(sequence))
sequence


519


'AGCTCTGAGAGAGGAGCCCAGCCCTGGGATTTTCAGGTGTTTTCATTTGGTGATCAGGACTGAACAGAGAGAACTCACCATGGAGTTTGGGCTGAGCTGGCTTTTTCTTGTGGCTATTTTAAAAGGTGTCCAGTGTGAGGTGCAGCTGTTGGAGTCTGGGGGAGGCTTGGTACAGCCTGGGGGGTCCCTGAGACTCTCCTGTGCAGCCTCTGGATTCACCTTTAGCAGCTATGCCATGAGCTGGGTCCGCCAGGCTCCAGGGAAGGGGCTGGAGTGGGTCTCAGCTATTAGTGGTAGTGGTGGTAGCACATACTACGCAGACTCCGTGAAGGGCCGGTTCACCATCTCCAGAGACAATTCCAAGAACACGCTGTATCTGCAAATGAACAGCCTGAGAGCCGAGGACACGGCCGTATATTACTGTGCGCGAAATTACTATGATAGTAGTGGTTATATTACTAACTTTGACTACTGGGGCCAGGGAACCCTGGTCACCGTCTCCTCAGCACCCACCAAGGC'

In [ ]:
sequence_alignment = heavy_sequence_df['sequence_alignment'][0]
print(len(sequence_alignment))
sequence_alignment

370


'GAGGTGCAGCTGTTGGAGTCTGGGGGAGGCTTGGTACAGCCTGGGGGGTCCCTGAGACTCTCCTGTGCAGCCTCTGGATTCACCTTTAGCAGCTATGCCATGAGCTGGGTCCGCCAGGCTCCAGGGAAGGGGCTGGAGTGGGTCTCAGCTATTAGTGGTAGTGGTGGTAGCACATACTACGCAGACTCCGTGAAGGGCCGGTTCACCATCTCCAGAGACAATTCCAAGAACACGCTGTATCTGCAAATGAACAGCCTGAGAGCCGAGGACACGGCCGTATATTACTGTGCGCGAAATTACTATGATAGTAGTGGTTATATTACTAACTTTGACTACTGGGGCCAGGGAACCCTGGTCACCGTCTCCTCAG'

In [ ]:
aligner = PairwiseAligner()
aligner.mode = 'local'
# Allow insertions at the 5' end of the target sequence without penalty
aligner.open_gap_score = -2.0        # Penalty for opening a gap
aligner.extend_gap_score = -0.5      # Penalty for extending a gap

alignments = aligner.align(sequence, sequence_alignment)
print(alignments[0])
print(alignments[0].score)

target          136 GAGGTGCAGCTGTTGGAGTCTGGGGGAGGCTTGGTACAGCCTGGGGGGTCCCTGAGACTC
                  0 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
query             0 GAGGTGCAGCTGTTGGAGTCTGGGGGAGGCTTGGTACAGCCTGGGGGGTCCCTGAGACTC

target          196 TCCTGTGCAGCCTCTGGATTCACCTTTAGCAGCTATGCCATGAGCTGGGTCCGCCAGGCT
                 60 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
query            60 TCCTGTGCAGCCTCTGGATTCACCTTTAGCAGCTATGCCATGAGCTGGGTCCGCCAGGCT

target          256 CCAGGGAAGGGGCTGGAGTGGGTCTCAGCTATTAGTGGTAGTGGTGGTAGCACATACTAC
                120 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
query           120 CCAGGGAAGGGGCTGGAGTGGGTCTCAGCTATTAGTGGTAGTGGTGGTAGCACATACTAC

target          316 GCAGACTCCGTGAAGGGCCGGTTCACCATCTCCAGAGACAATTCCAAGAACACGCTGTAT
                180 ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
query           180 GCAGACTCCGTGAAGGGCCGGTTCACCATCTCCAGAGACAATTCCAAGAACACGCTGTAT

target          376 CTGC

Sequence_alignments from IgBLAST are indeed a subsequence of the intiial Sequence column. Raw sequences may include leader peptides not included in the final form of the variable chain.